### PR TITLE
[INLONG-9904][Audit] SDK support checkpoint feature

### LINKED ISSUE #9904: [Feature][Audit] SDK supports checkpoint feature
### Description

SDK supports checkpoint feature

### Use case

SDK supports checkpoint feature

### Are you willing to submit PR?

- [X] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by doleyzi**
### Prepare a Pull Request
- Fixes #9904

### Motivation

*Flink checkpoint requires the audit SDK to support the isolation of audit data and ensure that each checkpoint cycle can be accurately reconciled.*

### Modifications

*The audit SDK supports checkpoint by adding resource isolation logic.*

### CODE REVIEW

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/send/SenderManager.java [hunk 1]
```diff
@@ -286,6 +269,9 @@ public void onMessageReceived(ChannelHandlerContext ctx, byte[] msg) {
             if (data == null) {
                 LOG.error("Can not find the request id onMessageReceived {},message: {}",
                         requestId, baseCommand.getAuditReply().getMessage());
+                for (Map.Entry<Long, AuditData> entry : this.dataMap.entrySet()) {
```
Comment (baomingyu) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/send/SenderManager.java:272]: It is better ?
if(LOG.isDebug()) {
    for ?
}

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java [hunk 2]
```diff
@@ -54,19 +58,36 @@ public class AuditReporterImpl implements Serializable {
...
+
+    public void setFlushStatThreshold(int flushStatThreshold) {
+        this.flushStatThreshold = flushStatThreshold;
+    }
+
+    public void setAutoFlush(boolean autoFlush) {
+        this.autoFlush = autoFlush;
+    }
+
+    private boolean autoFlush = true;
```
Comment (luchunliang) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:90]: Member variables need to be declared before all member methods.

Comment (doleyzi) [reply inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java]: resolved

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java [hunk 3]
```diff
@@ -198,50 +222,201 @@ public void add(int auditID, String auditTag, String inlongGroupID, String inlon
...
 
     /**
      * Add audit info by key.
      */
-    private void addByKey(String key, long count, long size, long delayTime) {
-        if (countMap.get(key) == null) {
-            countMap.put(key, new StatInfo(0L, 0L, 0L));
+    private void addByKey(long isolateKey, String statKey, long count, long size, long delayTime) {
+        if (null == this.preStatMap.get(isolateKey)) {
+            this.preStatMap.putIfAbsent(isolateKey, new ConcurrentHashMap<>());
```
Comment (baomingyu) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:242]: 'if' is not neccessary  check while use putIfAbsent function.

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java [hunk 4]
```diff
@@ -198,50 +222,201 @@ public void add(int auditID, String auditTag, String inlongGroupID, String inlon
...
         }
-        countMap.get(key).count.addAndGet(count);
-        countMap.get(key).size.addAndGet(size);
-        countMap.get(key).delay.addAndGet(delayTime);
+        ConcurrentHashMap<String, StatInfo> statMap = this.preStatMap.get(isolateKey);
+        if (null == statMap.get(statKey)) {
+            statMap.putIfAbsent(statKey, new StatInfo(0L, 0L, 0L));
```
Comment (baomingyu) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:246]: 'if' is not neccessary  check while use putIfAbsent function.

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java [hunk 5]
```diff
@@ -198,50 +222,201 @@ public void add(int auditID, String auditTag, String inlongGroupID, String inlon
...
-    public synchronized void send() {
+    public synchronized void flush() {
+        flush(DEFAULT_AUDIT_VERSION);
+    }
+
+    /**
+     * Flush audit data
+     */
+    public synchronized void flush(long isolateKey) {
+        if (flushTime.putIfAbsent(isolateKey, Calendar.getInstance().getTimeInMillis()) != null
```
Comment (luchunliang) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:265]: "Calendar.getInstance().getTimeInMillis()" is same with "System.currentTimeMillis". They both get TimeZone +00:00.

**Review by baomingyu — CHANGES_REQUESTED** [review]

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java [hunk 6]
```diff
@@ -198,50 +219,201 @@ public void add(int auditID, String auditTag, String inlongGroupID, String inlon
...
+     * Summary expired stat map
+     */
+    private void summaryExpiredStatMap(long isolateKey) {
+        Iterator<Map.Entry<Long, ConcurrentHashMap<String, StatInfo>>> iterator =
+                this.expiredStatMap.entrySet().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<Long, ConcurrentHashMap<String, StatInfo>> entry = iterator.next();
+            if (entry.getValue().isEmpty()) {
+                LOGGER.info("Remove the key of expired stat map: {},isolate key: {} ", entry.getKey(), isolateKey);
+                iterator.remove();
```
Comment (luchunliang) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:341]: It has synchronization issues between "entry.getValue().isEmpty()" and "iterator.remove()".

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java (hunk 3 above)
Comment (doleyzi) [reply inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:242]: The addKey function is a high-end function and will be called frequently, but put is a low-frequency operation. Therefore, using get, the performance consumption is less than putIfAbsent.

**Review by justinwwhuang — APPROVED** [review]

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java [hunk 7]
```diff
@@ -54,19 +58,33 @@ public class AuditReporterImpl implements Serializable {
...
     private SenderManager manager;
+    private AtomicInteger flushStat = new AtomicInteger(0);
 
     private final ScheduledExecutorService timeoutExecutor = Executors.newSingleThreadScheduledExecutor();
     private AuditConfig auditConfig = null;
     private SocketAddressListLoader loader = null;
+    private static final long DEFAULT_ISOLATE_KEY = 0;
+    private int flushStatThreshold = 100;
+    private boolean autoFlush = true;
+    public void setFlushStatThreshold(int flushStatThreshold) {
```
Comment (vernedeng) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:82]: Better add one blank line between variables and methods

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java [hunk 8]
```diff
@@ -54,19 +58,33 @@ public class AuditReporterImpl implements Serializable {
...
+        this.flushStatThreshold = flushStatThreshold;
+    }
+    public void setAutoFlush(boolean autoFlush) {
```
Comment (vernedeng) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:85]: ditto

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java (hunk 3 above)
Comment (vernedeng) [reply inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:242]: use computeIfAbsent

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java (hunk 4 above)
Comment (vernedeng) [reply inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:246]: ditto

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java [hunk 9]
```diff
@@ -198,50 +219,201 @@ public void add(int auditID, String auditTag, String inlongGroupID, String inlon
...
     }
 
     /**
-     * Send audit data
+     * Flush audit data
      */
-    public synchronized void send() {
+    public synchronized void flush(long isolateKey) {
+        if (flushTime.putIfAbsent(isolateKey, System.currentTimeMillis()) != null
+                || flushStat.addAndGet(1) > flushStatThreshold) {
```
Comment (vernedeng) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:266]: maybe is  <  ?

**Comment by EMsnap** [comment]
Could u please add a example or description on how to use this feature in checkpoint sence ?

**Review by vernedeng — DISMISSED** [review]
LGTM

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java [hunk 10]
```diff
@@ -54,19 +58,33 @@ public class AuditReporterImpl implements Serializable {
...
     private int packageId = 1;
     private int dataId = 0;
     private boolean initialized = false;
     private SenderManager manager;
+    private AtomicInteger flushStat = new AtomicInteger(0);
 
     private final ScheduledExecutorService timeoutExecutor = Executors.newSingleThreadScheduledExecutor();
     private AuditConfig auditConfig = null;
     private SocketAddressListLoader loader = null;
+    private static final long DEFAULT_ISOLATE_KEY = 0;
```
Comment (EMsnap) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:79]: And please add note for the parameter  isolateKey

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/util/AuditDimensions.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.inlong.audit.util;
+
+public class AuditDimensions {
```
Comment (EMsnap) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/util/AuditDimensions.java:20]: add note for new class

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java [hunk 12]
```diff
@@ -198,50 +219,201 @@ public void add(int auditID, String auditTag, String inlongGroupID, String inlon
         keyJoiner.add(String.valueOf(auditID));
         keyJoiner.add(auditTag);
         keyJoiner.add(String.valueOf(auditVersion));
-        addByKey(keyJoiner.toString(), count, size, delayTime);
+        addByKey(DEFAULT_ISOLATE_KEY, keyJoiner.toString(), count, size, delayTime);
+    }
+
+    public void add(AuditDimensions dimensions, AuditValues values) {
```
Comment (EMsnap) [inline inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:225]: method never been used?

File: inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java (hunk 4 above)
Comment (doleyzi) [reply inlong-audit/audit-sdk/src/main/java/org/apache/inlong/audit/AuditReporterImpl.java:246]: This function is a high-end function and will be called frequently, but put is a low-frequency operation. Therefore, using get, the performance consumption is less than putIfAbsent.

**Review by baomingyu — APPROVED** [review]

**Review by luchunliang — APPROVED** [review]

**Review by justinwwhuang — APPROVED** [review]
LGTM

**Review by dockerzhang — APPROVED** [review]

**Pull request merged**